File: llm_harness/stats/llm/llm_stats_stages/match_stage.py
"""Match stage for native Python LLM stats."""

from __future__ import annotations

from typing import Any

from ..matcher import get_scraper_fallback_match_diagnostics
from ..shared import as_record, model_slug_from_model_id, normalize_provider_model_id
from .types import LlmSourceData


def _has_token(model_id: str, token: str) -> bool:
    return token in model_id


def _canonical_model_id(model_id: Any, provider_id: Any, fallback_model_id: Any) -> str | None:
    if isinstance(model_id, str) and "/" in model_id:
        return model_id
    if isinstance(provider_id, str) and isinstance(model_id, str):
        return f"{provider_id}/{model_id}"
    if isinstance(provider_id, str) and isinstance(fallback_model_id, str):
        return f"{provider_id}/{fallback_model_id}"
    return model_id if isinstance(model_id, str) else None


def _has_variant_conflict(artificial_analysis_slug: str, matched_model_id: str, matcher_options: dict[str, Any]) -> bool:
    aa = normalize_provider_model_id(artificial_analysis_slug)
    matched = normalize_provider_model_id(matched_model_id)
    return any(_has_token(aa, token) != _has_token(matched, token) for token in matcher_options.get("variant_tokens", []))


def _build_matched_row_from_scraped_model(
    scraped_model: dict[str, Any],
    api_model: dict[str, Any] | None,
    matched_model_id: str,
    models_dev_by_id: dict[str, dict[str, Any]],
) -> dict[str, Any]:
    artificial_analysis_model_id = scraped_model.get("model_id") if isinstance(scraped_model.get("model_id"), str) else None
    artificial_analysis_slug = model_slug_from_model_id(artificial_analysis_model_id)
    evaluations = as_record(scraped_model.get("evaluations"))
    intelligence = as_record(scraped_model.get("intelligence"))
    intelligence_index_cost = as_record(scraped_model.get("intelligence_index_cost"))
    logo = scraped_model.get("logo") if isinstance(scraped_model.get("logo"), str) else None
    api_evaluations = as_record((api_model or {}).get("evaluations"))
    api_intelligence = as_record((api_model or {}).get("intelligence"))
    api_intelligence_index_cost = as_record((api_model or {}).get("intelligence_index_cost"))
    matched_models_dev = models_dev_by_id.get(matched_model_id)
    matched_model_fields = as_record(as_record(matched_models_dev).get("model"))
    canonical_id = _canonical_model_id(
        matched_model_fields.get("id") or matched_model_id,
        as_record(matched_models_dev).get("provider_id"),
        as_record(matched_models_dev).get("model_id"),
    )
    row = {key: value for key, value in matched_model_fields.items() if key not in {"id", "name", "family", "model_id", "slug"}}
    return {
        "id": canonical_id,
        "provider_id": as_record(matched_models_dev).get("provider_id"),
        "openrouter_id": matched_model_fields.get("id"),
        "name": matched_model_fields.get("name") if isinstance(matched_model_fields.get("name"), str) else artificial_analysis_model_id,
        "aa_id": artificial_analysis_model_id,
        "aa_slug": artificial_analysis_slug,
        "family": matched_model_fields.get("family"),
        "logo": logo,
        **row,
        "evaluations": {**api_evaluations, **evaluations},
        "intelligence": {**api_intelligence, **intelligence},
        "intelligence_index_cost": {**api_intelligence_index_cost, **intelligence_index_cost},
    }


def build_matched_payload(
    source_data: LlmSourceData,
    matcher_options: dict[str, Any] | None = None,
) -> dict[str, Any]:
    matcher_options = matcher_options or {"variant_tokens": ["flash-lite", "flash", "pro", "nano", "mini", "lite"]}
    models_dev_models = source_data["preferred_models_dev_models"]
    diagnostics = get_scraper_fallback_match_diagnostics(
        {
            "scraped_rows": source_data["scraped_rows"],
            "models_dev_models": models_dev_models,
        }
    )
    rows = []
    for matched_model in diagnostics.get("models", []):
        matched_model_id = as_record(matched_model.get("best_match")).get("model_id")
        if not isinstance(matched_model_id, str) or not matched_model_id:
            continue
        if _has_variant_conflict(matched_model["artificial_analysis_slug"], matched_model_id, matcher_options):
            continue
        scraped_slug = matched_model["artificial_analysis_slug"]
        scraped_model = source_data["scraped_by_slug"].get(scraped_slug)
        if not scraped_model:
            continue
        rows.append(
            _build_matched_row_from_scraped_model(
                scraped_model,
                source_data["api_by_slug"].get(scraped_slug),
                matched_model_id,
                source_data["models_dev_by_id"],
            )
        )
    return {
        "artificial_analysis_fetched_at_epoch_seconds": source_data["artificial_analysis_payload"].get("fetched_at_epoch_seconds"),
        "models_dev_fetched_at_epoch_seconds": source_data["models_dev_payload"].get("fetched_at_epoch_seconds"),
        "models": rows,
    }

File: llm-harness-js/src/stats/llm/llm-stats/match-stage.ts
import { getScraperFallbackMatchDiagnostics } from "../matcher.js";
import {
  asRecord,
  modelSlugFromModelId,
  normalizeProviderModelId,
  type JsonObject,
} from "../shared.js";

import {
  type ModelsDevModel,
  type ScrapedEvalModel,
  type SourceData,
} from "./types.js";

const MODEL_VARIANT_TOKENS = [
  "flash-lite",
  "flash",
  "pro",
  "nano",
  "mini",
  "lite",
] as const;

function hasToken(id: string, token: (typeof MODEL_VARIANT_TOKENS)[number]) {
  return id.includes(token);
}

function canonicalModelId(
  modelId: unknown,
  providerId: unknown,
  fallbackModelId: unknown,
): string | null {
  if (typeof modelId === "string" && modelId.includes("/")) {
    return modelId;
  }
  if (typeof providerId === "string" && typeof modelId === "string") {
    return `${providerId}/${modelId}`;
  }
  if (typeof providerId === "string" && typeof fallbackModelId === "string") {
    return `${providerId}/${fallbackModelId}`;
  }
  return typeof modelId === "string" ? modelId : null;
}

function hasVariantConflict(
  artificialAnalysisSlug: string,
  matchedModelId: string,
): boolean {
  const aa = normalizeProviderModelId(artificialAnalysisSlug);
  const matched = normalizeProviderModelId(matchedModelId);
  return MODEL_VARIANT_TOKENS.some(
    (token) => hasToken(aa, token) !== hasToken(matched, token),
  );
}

function buildMatchedRowFromScrapedModel(
  scrapedModel: ScrapedEvalModel,
  matchedModelId: string,
  modelsDevById: Map<string, ModelsDevModel>,
): Record<string, unknown> {
  const artificialAnalysisModelId =
    typeof scrapedModel.model_id === "string" ? scrapedModel.model_id : null;
  const artificialAnalysisSlug = modelSlugFromModelId(
    artificialAnalysisModelId,
  );
  const evaluations = asRecord(scrapedModel.evaluations);
  const intelligence = asRecord(scrapedModel.intelligence);
  const intelligenceIndexCost = asRecord(scrapedModel.intelligence_index_cost);
  const logo = typeof scrapedModel.logo === "string" ? scrapedModel.logo : null;
  const matchedModelsDev = modelsDevById.get(matchedModelId) ?? null;
  const matchedModelFields = asRecord(matchedModelsDev?.model);
  const canonicalId = canonicalModelId(
    matchedModelsDev?.model?.id ?? matchedModelId,
    matchedModelsDev?.provider_id,
    matchedModelsDev?.model_id,
  );
  const {
    id: _matchedId,
    name: _matchedName,
    family: matchedFamily,
    model_id: _matchedModelId,
    slug: _matchedSlug,
    ...matchedModelFieldsWithoutIdFamilyAndModelRefs
  } = matchedModelFields;

  return {
    id: canonicalId,
    provider_id: matchedModelsDev?.provider_id ?? null,
    openrouter_id: matchedModelsDev?.model?.id ?? null,
    name:
      typeof matchedModelsDev?.model?.name === "string"
        ? matchedModelsDev.model.name
        : artificialAnalysisModelId,
    aa_id: artificialAnalysisModelId,
    aa_slug: artificialAnalysisSlug,
    family: matchedFamily,
    logo,
    ...matchedModelFieldsWithoutIdFamilyAndModelRefs,
    evaluations,
    intelligence,
    intelligence_index_cost: intelligenceIndexCost,
  };
}

export async function buildMatchedRows(
  sourceData: SourceData,
): Promise<Record<string, unknown>[]> {
  const fallbackDiagnostics = await getScraperFallbackMatchDiagnostics({
    scrapedRows: sourceData.scrapedRows,
    modelsDevModels: sourceData.preferredModelsDevModels,
  });

  return fallbackDiagnostics.models
    .map((matchedModel) => {
      const matchedModelId = matchedModel.best_match?.model_id;
      if (typeof matchedModelId !== "string" || matchedModelId.length === 0) {
        return null;
      }
      if (
        hasVariantConflict(
          matchedModel.artificial_analysis_slug,
          matchedModelId,
        )
      ) {
        return null;
      }
      const scrapedModel = sourceData.scrapedBySlug.get(
        matchedModel.artificial_analysis_slug,
      );
      if (!scrapedModel) {
        return null;
      }
      return buildMatchedRowFromScrapedModel(
        scrapedModel,
        matchedModelId,
        sourceData.modelsDevById,
      );
    })
    .filter((row): row is Record<string, unknown> => row != null);
}
